fix and improve AppImage

Samueru-sama opened the pull request:
I have not tested beyond launching the AppImage on alpine linux, which is usually a good enough first test. 

@lestcape and @Dangonizer [artifact link](https://github.com/Samueru-sama/RMG/actions/runs/12856278768/artifacts/2453287353)

Let me know if there are any issues.

---------------------------------------------

edit: Now the AppImage follows the system theme if you use qt6ct for that. 

[It also can load separate stylesheet files](https://github.com/Samueru-sama/RMG/blob/appimage-anylinux/Package/AppImage/Create.sh#L46-L47) by either placing and renaming the file with the name of the appimage + `.stylesheet` or directly setting the env variable to the location of the file.


edit2: closes https://github.com/pkgforge-dev/Anylinux-AppImages/issues/3 #323

Rosalie241 commented:
I'm not a huge fan of the big size increase, it goes from 80MB -> 140MB which is a *big* jump.

I've also tried it for a little bit and it has a few issues.

1) Netplay, it fails to initialize TLS, you should be able to reproduce this by opening the Netplay -> Create Session dialog, you can view the qt log by setting `RMG_SHOW_DEBUG_QT_MESSAGES=1`:
```
[QT WARNING] No TLS backend is available
[QT WARNING] No functional TLS backend was found
[QT WARNING] QSslSocket::connectToHostEncrypted: TLS initialization failed
```

2) Vulkan video plugins do not work, you can try this by going to Settings -> Plugins and changing the RSP and Video plugin to paraLLEl and booting a game, and then you can view the error log in View -> Log:
```
[GFX]   [ERROR]   Failed to create Vulkan instance.
[GFX]   [ERROR]   Failed to create Vulkan instance.
[GFX]   [ERROR]   Could not start GFX plugin
```
From a quick look, it seems to be missing some vulkan extensions, on a working AppImage you see the following:
```
[GFX]   [INFO]    Layer count: 9
[GFX]   [INFO]    Found layer: VK_LAYER_VALVE_steam_overlay_32.
[GFX]   [INFO]    Found layer: VK_LAYER_VALVE_steam_overlay_64.
[GFX]   [INFO]    Found layer: VK_LAYER_VALVE_steam_fossilize_32.
[GFX]   [INFO]    Found layer: VK_LAYER_VALVE_steam_fossilize_64.
[GFX]   [INFO]    Found layer: VK_LAYER_MANGOHUD_overlay_x86.
[GFX]   [INFO]    Found layer: VK_LAYER_MANGOAPP_overlay.
[GFX]   [INFO]    Found layer: VK_LAYER_MANGOHUD_overlay_x86_64.
[GFX]   [INFO]    Found layer: VK_LAYER_MANGOAPP_overlay.
[GFX]   [INFO]    Found layer: VK_LAYER_MESA_device_select.
[GFX]   [INFO]    Enabling instance extension: VK_KHR_device_group_creation.
[GFX]   [INFO]    Enabling instance extension: VK_KHR_display.
[GFX]   [INFO]    Enabling instance extension: VK_KHR_external_fence_capabilities.
[GFX]   [INFO]    Enabling instance extension: VK_KHR_external_memory_capabilities.
[GFX]   [INFO]    Enabling instance extension: VK_KHR_external_semaphore_capabilities.
[GFX]   [INFO]    Enabling instance extension: VK_KHR_get_display_properties2.
[GFX]   [INFO]    Enabling instance extension: VK_KHR_get_physical_device_properties2.
[GFX]   [INFO]    Enabling instance extension: VK_KHR_get_surface_capabilities2.
[GFX]   [INFO]    Enabling instance extension: VK_KHR_surface.
[GFX]   [INFO]    Enabling instance extension: VK_KHR_surface_protected_capabilities.
[GFX]   [INFO]    Enabling instance extension: VK_KHR_wayland_surface.
[GFX]   [INFO]    Enabling instance extension: VK_KHR_xcb_surface.
[GFX]   [INFO]    Enabling instance extension: VK_KHR_xlib_surface.
[GFX]   [INFO]    Enabling instance extension: VK_EXT_acquire_drm_display.
[GFX]   [INFO]    Enabling instance extension: VK_EXT_acquire_xlib_display.
[GFX]   [INFO]    Enabling instance extension: VK_EXT_debug_report.
[GFX]   [INFO]    Enabling instance extension: VK_EXT_debug_utils.
[GFX]   [INFO]    Enabling instance extension: VK_EXT_direct_mode_display.
[GFX]   [INFO]    Enabling instance extension: VK_EXT_headless_surface.
[GFX]   [INFO]    Enabling instance extension: VK_EXT_surface_maintenance1.
[GFX]   [INFO]    Enabling instance extension: VK_EXT_swapchain_colorspace.
[GFX]   [INFO]    Enabling instance extension: VK_EXT_display_surface_counter.
[GFX]   [INFO]    Enabling instance extension: VK_KHR_portability_enumeration.
[GFX]   [INFO]    Enabling instance extension: VK_LUNARG_direct_driver_loading.
[GFX]   [INFO]    Enabling instance extension: VK_EXT_debug_utils.
[GFX]   [INFO]    Enabling instance extension: VK_KHR_get_surface_capabilities2.
[GFX]   [INFO]    Enabling instance extension: VK_EXT_surface_maintenance1.
[GFX]   [INFO]    Enabling instance extension: VK_EXT_swapchain_colorspace.
```
while on a non-working AppImage like the one you have, you can see:
```
[GFX]   [INFO]    Layer count: 9
[GFX]   [INFO]    Found layer: VK_LAYER_VALVE_steam_overlay_32.
[GFX]   [INFO]    Found layer: VK_LAYER_VALVE_steam_overlay_64.
[GFX]   [INFO]    Found layer: VK_LAYER_VALVE_steam_fossilize_32.
[GFX]   [INFO]    Found layer: VK_LAYER_VALVE_steam_fossilize_64.
[GFX]   [INFO]    Found layer: VK_LAYER_MANGOHUD_overlay_x86.
[GFX]   [INFO]    Found layer: VK_LAYER_MANGOAPP_overlay.
[GFX]   [INFO]    Found layer: VK_LAYER_MANGOHUD_overlay_x86_64.
[GFX]   [INFO]    Found layer: VK_LAYER_MANGOAPP_overlay.
[GFX]   [INFO]    Found layer: VK_LAYER_MESA_device_select.
[GFX]   [INFO]    Enabling instance extension: VK_EXT_debug_report.
[GFX]   [INFO]    Enabling instance extension: VK_EXT_debug_utils.
[GFX]   [INFO]    Enabling instance extension: VK_KHR_portability_enumeration.
[GFX]   [INFO]    Enabling instance extension: VK_LUNARG_direct_driver_loading.
[GFX]   [INFO]    Enabling instance extension: VK_EXT_debug_utils.
```

I also have another question, considering you're shipping `libGL`, `libEGL` and such, do nonfree drivers like the nVidia drivers still work with this?

Samueru-sama commented:
> I'm not a huge fan of the big size increase, it goes from 80MB -> 140MB which is a _big_ jump.

The current AppImage does not launch on my system nor on the ones of the users that reported the issue.

>Netplay, it fails to initialize TLS, you should be able to reproduce this by opening the Netplay -> Create Session dialog, you can view the qt log by setting RMG_SHOW_DEBUG_QT_MESSAGES=1:

mmmm, that's odd I guess the libraries needed for this are only dlopened when such function is used. 

`lib4bin` uses a mode that launches the app with strace to get all the libraries it needs, however I think it is failing to work, likely because the app refuses to run properly when it detects that it is ran as sudo in the CI or something similar. 

>Vulkan video plugins do not work, you can try this by going to Settings -> Plugins and changing the RSP and Video plugin to paraLLEl and booting a game, and then you can view the error log in View -> Log:

I think I only bundled opengl, wasn't aware this used vulkan as well 👀

>I also have another question, considering you're shipping libGL, libEGL and such, do nonfree drivers like the nvidia drivers still work with this?

[Yes](https://github.com/probonopd/PrusaSlicer/pull/5). Actually that's the reason because I switched to building on 24.04, because otherwise there is a bug on old mesa that prevents the proprietary nvidia driver from being used.

sharun still adds the host [glvd](https://github.com/Samueru-sama/sharun/blob/main/src/main.rs#L522) to `__EGL_VENDOR_LIBRARY_DIRS` to let the proprietary nvidia driver be used. 

It also does a similar thing for vulkan just in case.

Rosalie241 commented:
> The current AppImage does not launch on my system nor on the ones of the users that reported the issue.

That'd odd, have you tried debugging why it doesn't launch? Maybe try setting `RMG_SHOW_DEBUG_QT_MESSAGES=1` or using strace?

> sharun still adds the host [glvd](https://github.com/Samueru-sama/sharun/blob/main/src/main.rs#L522) to __EGL_VENDOR_LIBRARY_DIRS to let the proprietary nvidia driver be used.

That's good

Samueru-sama commented:
>Netplay, it fails to initialize TLS, you should be able to reproduce this by opening the Netplay -> Create Session dialog, you can view the qt log by setting RMG_SHOW_DEBUG_QT_MESSAGES=1:

Alright it is very likely that this is because the qt tls plugin wasn't added at all, likely indicating that the app isn't running in the CI, hence none of the dlopened libs are being picked up. 

If this still fails I will  start doing proper debugging with strace.

Rosalie241 commented:
I'd rather have you look into *why* the existing AppImage doesn't work and attempt to fix that instead of switching to an entirely different system without a good reason, because I'm currently unconvinced that the new system is better than the existing one, and not knowing why the existing one doesn't work isn't very helpful either.

Samueru-sama commented:
> I'd rather have you look into _why_ the existing AppImage doesn't work and attempt to fix that instead of switching to an entirely different system without a good reason, because I'm currently unconvinced that the new system is better than the existing one, and not knowing why the existing one doesn't work isn't very helpful either.

The existing appimage isn't great, relying on the host libraries is just not reliable you will often see issues of the AppImage not working on certain users. **that's the good reason**.

>I'm currently unconvinced that the new system is better than the existing one

Because I forgot to add vulkan and the tls plugin? Note I'm assuming that fixed those issues, if it didn't I'm very sorry.

Rosalie241 commented:
> The existing appimage isn't great, relying on the host libraries is just not reliable you will often see issues of the AppImage not working on certain users. that's the good reason.

That's a fair point, sure.

> EDIT: Updated artifact link, also putting it [here](https://github.com/Samueru-sama/RMG/actions/runs/12856278768/artifacts/2453287353)

I can sadly confirm that both issues remain:
```
[QT WARNING] The backend "cert-only" does not support QSslKey
[QT WARNING] Active TLS backend does not support key creation
[QT WARNING] The backend "cert-only" does not support QSslKey
[QT WARNING] Active TLS backend does not support key creation
[QT WARNING] The backend "cert-only" does not support QSslSocket
[QT WARNING] The backend named "cert-only" does not support TLS
[QT WARNING] QSslSocket::connectToHostEncrypted: TLS initialization failed
```

And the vulkan video plugin has the same issue, it seems like not all the required instance extensions are being exposed.

Samueru-sama commented:
> > The existing appimage isn't great, relying on the host libraries is just not reliable you will often see issues of the AppImage not working on certain users. that's the good reason.
> 
> That's a fair point, sure.
> 
> > EDIT: Updated artifact link, also putting it [here](https://github.com/Samueru-sama/RMG/actions/runs/12856278768/artifacts/2453287353)
> 
> I can sadly confirm that both issues remain:
> 
> ```
> [QT WARNING] Incompatible version of OpenSSL (built with OpenSSL >= 3.x, runtime version is < 3.x)
> [QT WARNING] The backend "cert-only" does not support QSslKey
> [QT WARNING] Active TLS backend does not support key creation
> [QT WARNING] The backend "cert-only" does not support QSslKey
> [QT WARNING] Active TLS backend does not support key creation
> [QT WARNING] The backend "cert-only" does not support QSslSocket
> [QT WARNING] The backend named "cert-only" does not support TLS
> [QT WARNING] QSslSocket::connectToHostEncrypted: TLS initialization failed
> ```
> 
> And the vulkan video plugin has the same issue, it seems like not all the required instance extensions are being exposed.

Alright I'm very sorry, gotta properly setup tests in that case.

Samueru-sama commented:
Just in case @Rosalie241 go with the rest of your day and ignore everything I do here until I notify everything is ready, I will make sure everything is fixed. 

I should get all of this sorted out in a few days at most (gotta go just now lol) and I also have something in mind that should reduce the size of the AppImage by about 20 MiB.

lestcape commented:
The app image works for me and I don't see any issues, but I don't really have many complicated use cases.

@Rosalie241, if some dependency of an app image is not built in and that dependency changes in the next release or is missing in a future state, it's impossible for the app image to work. So the more dependencies are built in, the less chances there are for the app image to fail in the future or work in an unplanned environment, but of course more built in libraries require more size of the image. So size can't be seen as a problem, but as the solution to the problem, since the main objective of an appimage is precisely to run without attachment to a specific environment

Having an appimage that works only on the system where it was generated or on systems almost identical to it, is practically the same as compiling for a specific environment and therefore the essence of what it means to have an appimage is lost.

I would like that 20 years later, when nothing resembles what we have today, the appimage that was created today still works. That is the true utility of having a sufficiently robust appimage. But of course, the appimage should be complete and functional today so that it will be functional in the future. 

Maybe what you should consider is whether you are going to generate an appimage in all releases or if, on the contrary, you are going to generate it only when there is a really noticeable change in your app. Is preferable to me 1 appimage per year that works for 20 years that 1 appimage per day that works for a week or only works on the distro where it was created.

By the way, I really liked your emulator. Especially the feature of being able to decide whether to let keyboard shortcuts repeat or not is something all emulators should have.

Samueru-sama commented:
Turns out vulkan drivers were not installed at all in the CI 🤦

Now I'm able to launch a title with vulkan. I can also go into netplay without having an error, although I don't see any session to join to, not sure if I need to configure this. 

Updated artifact link, [size is now 155 MiB](https://github.com/Samueru-sama/RMG/actions/runs/12857639203/artifacts/2453569453) which is very close to the dolphin emu appimage, after all vulkan, opengl and all of qt are bundle in both. 

I can reduce the final size to 140 MiB by using a different appimage runtime that uses dwarfs instead of squashfs, however this has the downside that it breaks everything that interacts with squash in appimages, most notably the AppImage thumbnailers that some users may want: 

![image](https://github.com/user-attachments/assets/b76a777d-5e36-426e-95ef-a166f827db08)

Some might suggest to build on alpine linux, but this will stop people from using the proprietary nvidia driver at all, since this is dlopened on the host means what we ship in the AppImage has to have glibc yes or yes. 

This means the only other method to reduce size is to build some dependencies statically, but I really don't know how to do that. 

I know it can reduce the size a lot, because I make this [AppImage of mpv](https://github.com/pkgforge-dev/mpv-AppImage) (bundles everything) and it used to be 80 MIB before I switched to their build script that automatically builds it linking ffmpeg statically, that reduced its size from **80 MiB to 22 MiB**.


-------------------------------------------------------

So to resume so far this PR: 

* Fixes the issue of the AppImage not opening, although I would want confirmation from @Dangonizer still. 
* Lets the AppImage follow the system theme if you have qt6ct, I guess more styles could be bundled but this means the size increases as well. 
* Adds delta updates with [AppImageUpdate](https://github.com/AppImageCommunity/AppImageUpdate). Although I haven't modified the CI to also upload the `.zsync` file yet. (Note that with the dwarfs runtime this would still work just in case). 
* Adds wayland support since now I'm bundling the wayland plugins. Hasn't been test it though.

Dangonizer commented:
> * Fixes the issue of the AppImage not opening, although I would want confirmation from @Dangonizer still.

Can confirm that it launches now. 👍

Rosalie241 commented:
I can confirm the issues have been fixed, thank you.

> Updated artifact link, [size is now 155 MiB](https://github.com/Samueru-sama/RMG/actions/runs/12857639203/artifacts/2453569453) which is very close to the dolphin emu appimage, after all vulkan, opengl and all of qt are bundle in both.

That's rather big but at this point I'll just live with it, sadly RMG is already quite large due to Qt6 and shipping vosk-api.

> I can reduce the final size to 140 MiB by using a different appimage runtime that uses dwarfs instead of squashfs, however this has the downside that it breaks everything that interacts with squash in appimages, most notably the AppImage thumbnailers that some users may want:

I think having a 'standard' runtime makes more sense.

> Some might suggest to build on alpine linux, but this will stop people from using the proprietary nvidia driver at all, since this is dlopened on the host means what we ship in the AppImage has to have glibc yes or yes.

I don't even know if RMG will build on alphine linux, though I'm a little curious about the size for it now, might experiment with that in the future but for now glibc is fine also due to the nvidia support.

> Lets the AppImage follow the system theme if you have qt6ct, I guess more styles could be bundled but this means the size increases as well.

I'd rather not ship this in the AppImage and have RMG handle it on it's own, RMG supports qt stylesheets by itself but it's an undocumented feature.

> Adds delta updates with [AppImageUpdate](https://github.com/AppImageCommunity/AppImageUpdate). Although I haven't modified the CI to also upload the .zsync file yet. (Note that with the dwarfs runtime this would still work just in case).

RMG handles updates for AppImage by itself, so the zsync stuff is unrequired.

> Adds wayland support since now I'm bundling the wayland plugins. Hasn't been test it though.

I can confirm this works fine on my system by setting `RMG_WAYLAND=1`, the wayland support is locked behind an environment variable because I've experienced a lot of issues on GNOME last time I tested it on that DE.

Samueru-sama commented:
>I don't even know if RMG will build on alphine linux, though I'm a little curious about the size for it now, might experiment with that in the future but for now glibc is fine also due to the nvidia support.

It usually doesn't reduce the size more than 20%, the only thing that can reduce at lot is static linking some dependencies.

[For example this is a Qt5 app that bundles everything](https://github.com/pkgforge-dev/SpeedCrunch-AppImage/releases/tag/continuous), the reason it is that small is because the developers released a portable build that linked Qt5 statically, so I'm pretty much only adding glibc and the dynamic linker to the appimage to make it work on any linux system. 

Of course it doesn't contain stuff like vulkan and opengl, but even with that the final size would be under 100 MiB. 

>I'd rather not ship this in the AppImage and have RMG handle it on it's own

So I remove `qt6ct` from the CI? Note that this is unlikely to reduce the size by more than 1 MIB. And it is a feature you normally get when it is a native distro package for example. just let me know anyway. 

I would not really want to remove [this other method](https://github.com/Samueru-sama/RMG/blob/appimage-anylinux/Package/AppImage/Create.sh#L49-L50) that I added to load stylesheets outside the application, basically it works by placing the stylesheet file next to the appimage with the name of the appimage + `.stylesheet` or directly setting the env variable on your shell profile to point to such file.

Rosalie241 commented:
> So I remove qt6ct from the CI? Note that this is unlikely to reduce the size by more than 1 MIB. And it is a feature you normally get when it is a native distro package for example. just let me know anyway.
> I would not really want to remove [this other method](https://github.com/Samueru-sama/RMG/blob/appimage-anylinux/Package/AppImage/Create.sh#L49-L50) that I added to load stylesheets outside the application, basically it works by placing the stylesheet file next to the appimage with the name of the appimage + .stylesheet or directly setting the env variable on your shell profile to point to such file.

Well alright, you can leave it in I suppose, I'll start reviewing your code changes now though.

Rosalie241 commented on Package/AppImage/Create.sh:
```diff
@@ -4,37 +4,65 @@ set -ex
-        -o "$script_dir/linuxdeploy-x86_64.AppImage"
-    chmod +x "$script_dir/linuxdeploy-x86_64.AppImage"
-fi
+cd "$bin_dir"
```
Can you try to remove the usage of `cd` in this script? you can use `$script_dir` to get the current directory.

Samueru-sama replied:
`cd` is needed because `lib4bin` creates the `AppDIr` structure inside the current working directory. I often just do `git clone && make && make install` to simplify things more since I just need to give `lib4bin` the path to the ELFs to bundle instead of having to change the install location when building.

Rosalie241 replied:
From a quick look at the `lib4bin` script, it seems to have a `-d, --dst-dir '/path'` commandline option, see https://github.com/VHSgunzo/sharun/blob/main/lib4bin#L41

Samueru-sama replied:
[It was added 2 months ago](https://github.com/VHSgunzo/sharun/commit/3ff9aef52796397bc6cdc6d414b48f23ef4238c4) 😭

Alright, fixing in that case.

Rosalie241 commented on Package/AppImage/Create.sh:
```diff
@@ -4,37 +4,65 @@ set -ex
-        -o "$script_dir/linuxdeploy-plugin-qt-x86_64.AppImage"
-    chmod +x "$script_dir/linuxdeploy-plugin-qt-x86_64.AppImage"
+if [ ! -f "./lib4bin" ]; then
+	wget --retry-connrefused --tries=30 "$LIB4BN" -O ./lib4bin
```
Can you use `curl` like this script had before? Having wget with 30 tries seems excessive for something that's mainly run in CI.

Samueru-sama replied:
Believe me the 30 times are needed 👀 github CI sometimes just dies for some reason with connection errors. 

The 30 times are done only if the connection fails. But sure no problem switching to curl.

Rosalie241 commented on Package/AppImage/Create.sh:
```diff
@@ -4,37 +4,65 @@ set -ex
-    "$script_dir/linuxdeploy-plugin-qt-x86_64.AppImage"
+# Prepare sharun
+./sharun -g
+echo '#!/usr/bin/env sh
```
Can you put this back in the `AppRun` file that you removed and copy that over instead of doing an echo/cat?

Samueru-sama replied:
I would need to modify the script to find the AppRun on the repo and put it in the AppDir. So ok.

Rosalie241 commented on Package/AppImage/Create.sh:
```diff
@@ -4,37 +4,65 @@ set -ex
+fi
+./appimagetool --comp zstd \
+	--mksquashfs-opt -Xcompression-level --mksquashfs-opt 22 \
+	-n -u "$UPINFO" "$PWD"/AppImage "$PWD"/RMG-"$VERSION"-anylinux-"$ARCH".AppImage
```
You can use `$script_dir` instead of `$PWD` here, I'd also prefer you didn't change the filename and restored it to what it was.

Samueru-sama replied:
I'm using `$PWD` because I'm already next to the `AppDir` because we had to `cd` into it to use `lib4bin`. 

The reason I use `$PWD` instead of just `./` is because I have had appimagetool bugged out when given a relative path.

Samueru-sama replied:
Note the AppImage spec suggests using `AppName-VERSION-ARCH` for the naming scheme of the AppImage, the `anylinux` is something I just add to indicate that this appimage works on even more systems than the usual. 

but sure going to pass`$OUTPUT` to appimagetool.

Rosalie241 commented:
Thank you for the PR :heart:

lestcape commented:
Thanks so much, for this @Rosalie241 and @Samueru-sama. We finally have a funtional portable version of a good N64 emulator!! Have a good day.

Samueru-sama commented:
@Rosalie241 One thing that can be done to reduce the size of the AppImage is to build a debloated version of llvm, which isn't as compilcated as building statically. 

One problem with bundling mesa is that the opengl part links to llvm, and `libLLVM.so` is a huge library that's used for a lot of stuff on linux, from compilers, Jit in browsers, and shader building, and we usually only need the later. On my Artixlinux system the lib 134 MiB. 

Distros also build the lib by default with support for ARM, AArch64, riscv, etc, etc. 

If the lib gets built with support for X86, AMDGPU with the flag `-DCMAKE_BUILD_TYPE=MinSizeRel` this cuts down the size of the lib to **66 MiB**, which translates to about 20 MIB saved in the final appimage. [1](https://github.com/pkgforge-dev/Cromite-AppImage/pull/13) [2](https://github.com/pkgforge-dev/Citron-AppImage/pull/2) [3](https://github.com/pkgforge-dev/PPSSPP-AppImage/pull/8).

The problem here is that the build would need to be switched to Archlinux and you would rely on the llvm package that I make, or build the debloated llvm which while it is a simple thing to do it takes hours in github actions. 

Anyways, just letting you know, if you are interested in these optimizations let me know.

Rosalie241 commented:
I personally think the way it currently is, is fine, yes it's quite a big AppImage, but using a custom LLVM build is rather annoying to rely on.
